### ISSUE TITLE
Investigate occasional failures of sync test with API rate limit breach on attempt to download a new note from a linked notebook

### ISSUE THREAD

**Issue by d1vanov** [d1vanov/libquentier#60]
One of tests for the synchronization logic implemented in #54 seems to occasionally fail and for some reason only on builds which run on launchpad build agents. I cannot reproduce the failure on any local machine of mine so we have a classic heisenbug here...

The tactics of nailing it down is going to be the following:
 * Add some kind of diagnostic output to print on test failure so that at least some bits of useful information can be extracted from the failing test's output
 * Use Lenin's squint i.e. stare at the source code of the sync algorithm as long as required to figure out where the error can be

**Comment by d1vanov** [comment]
Yay, managed to reproduce it on [Travis CI](https://travis-ci.org/d1vanov/libquentier/builds/393070906) and now the trace-level log of a failing test is available. So fixing this issue should now be just a matter of time for me to sit and examine this log carefully to understand the root cause of the issue.

**Comment by d1vanov** [comment]
Oh crap. Travis CI truncated the log after it reached 4 Mb :cry: I can't use it to diagnose the problem. Well, will try to see how would things be going on launchpad this time.

**Comment by d1vanov** [comment]
Managed to reproduce on launchpad, log is available, attaching it here just in case.
[buildlog_ubuntu-xenial-i386.libquentier_0.4.0-4_ppa_quentier_201806161648_ubuntu16.04.1_BUILDING.txt.gz](https://github.com/d1vanov/libquentier/files/2108362/buildlog_ubuntu-xenial-i386.libquentier_0.4.0-4_ppa_quentier_201806161648_ubuntu16.04.1_BUILDING.txt.gz)

**Comment by d1vanov** [comment]
And two more logs from launchpad, one from Qt4 build, another one - from Qt5 build.

[buildlog_ubuntu-artful-amd64.libquentier_0.4.0-3_ppa_quentier_201806161704_ubuntu17.10.1_BUILDING.txt.gz](https://github.com/d1vanov/libquentier/files/2108370/buildlog_ubuntu-artful-amd64.libquentier_0.4.0-3_ppa_quentier_201806161704_ubuntu17.10.1_BUILDING.txt.gz)

[buildlog_ubuntu-bionic-amd64.libquentier_0.4.0-3_ppa_quentier_201806161704_ubuntu18.04.1_BUILDING.txt.gz](https://github.com/d1vanov/libquentier/files/2108376/buildlog_ubuntu-bionic-amd64.libquentier_0.4.0-3_ppa_quentier_201806161704_ubuntu18.04.1_BUILDING.txt.gz)

**Comment by d1vanov** [comment]
It seems I finally managed to find the reason and fix this bug correctly. At least it no longer reproduces even on launchpad. In addition to fixing the logic causing test failures I also removed the necessity to have `LocalStorageManagerAsync` and `SynchronizationManager` in different threads of execution, now they can perfectly fine work within the same thread. I used this circumstance to simplify the setup of tests as sometimes they failed due to some problems caused by premature thread termination.

Finished after commit [175f4a9](https://github.com/d1vanov/libquentier/commit/175f4a91154d6ba54911fa8348c47d43b2bcd233) in branch [development](https://github.com/d1vanov/libquentier/commit/development).
